### PR TITLE
RFS (Responsive font size) implementation

### LINKED ISSUE #23053: Responsive font sizes
Currently there is no font-resizing implemented in Bootstrap. Font-resizing is needed to make large texts readable on all devices and to prevent longer words and [compounds](https://en.wikipedia.org/wiki/Compound_(linguistics)) to fall off screen. 

It's currently possible to alter the font-size of the html element with media queries (see [Responsive typography](https://getbootstrap.com/docs/4.1/content/typography/#responsive-typography)), but this method can cause the font-size of text to be too small on mobile devices and because the rem-value is overridden, users can't change their font-size in their browser anymore.

[RFS](https://github.com/MartijnCuppens/rfs) can solve this by because this mixin automatically calculates the appropriate font-size based on the width of the device. 

RFS is also available in Sass, Less, PostCSS and Stylus.

**Example problem with bigger font-sizes:** 
http://getbootstrap.com/docs/4.1/content/typography/#display-headings

**RFS solution:**
https://project-rfs.github.io/bootstrap/docs/4.1/content/typography/#display-headings

**Bootstrap implementation demo:** 
https://project-rfs.github.io/bootstrap/

**Github page:** 
https://github.com/project-rfs/rfs

**Pull request**
https://github.com/twbs/bootstrap/pull/23816

### PR DESCRIPTION
**Pull request by MartijnCuppens**
Fixes #23053

- RFS is disabled by default and can be switched on with `$enable-responsive-font-sizes`
- `font-size`-properties are switched to the `@include font-size()`-mixin. Stylelint prevents the usage of `font-size` property.
- Basic documentation added + link to github repo for further documentation.
- RFS is enabled to rescale font-sizes of titles on the docs page.

Demo with RFS enabled available here (this is not the default behaviour, but would be the behaviour if `$enable-responsive-font-sizes` was `true`): https://project-rfs.github.io/

TODO:

- [x] Remove fusv false warning workaround
- [x] Decide whether or not we're going to keep `$input-height-sm` and `$input-height-lg` > no
- [x] Test if we should increase the minimum font size to 1.25rem > yes

### CODE REVIEW

**Comment by MartijnCuppens** [comment]
@mdo: what do you think of this?

**Comment by gijsbotje** [comment]
isn't this the third time you made a pr for this?

**Comment by MartijnCuppens** [comment]
@gijsbotje, yes it is. The first PR contained generated css and had a lot of merge conflicts since the release of Bootstrap 4 beta. In the second PR the configuration was still in `px` (because this was how RFS was originally designed), but therefore a lot needed to be changed. Both PR were closed by me. 

This PR contains an updated version of RFS which allowed me to keep the current font-size configuration.

**Comment by mdo** [comment]
Thanks for keeping this up to date. I'm slating this for sure for 4.1, but I'll see if we should put it into 4.0 depending on a few other things.

**Comment by MartijnCuppens** [comment]
Update: It's now possible to disable responsive font-sizes for page by adding the `.disable-responsive-font-size` class to the body or html. This class can also be added to other elements.

**Comment by wolfy1339** [comment]
`$enable-hover-media-query` was not deprecated

**Comment by MartijnCuppens** [comment]
@wolfy1339: https://github.com/twbs/bootstrap/pull/25270/files#diff-d8ee409a461718bfb6240710c8c73382

**Comment by patrickhlauke** [comment]
This is all voodoo to me, but one request: make sure that even if using `vh`/`vw` etc, that there's a component for font sizing tied to something else like `px`/`rem`/`em`. Otherwise (when a font size is purely related to viewport dimensions and nothing else) you kill the user's ability to (full page) zoom on desktop and change the size the text is rendered at (as yes, the viewport dimensions change, but the font size would then change proportionally in the opposite direction, meaning the apparent font size is never actually altered).

**Comment by MartijnCuppens** [comment]
@patrickhlauke, It's possible to add the `disable-responsive-font-size` class to achieve this:
https://codepen.io/MartijnCuppens/pen/PEgNVM?editors=1100

**Comment by MartijnCuppens** [comment]
There are several ways to implement this depending on what should be the default value for `$rfs-class`.

In the current implementation I've set `$rfs-class: "disable";`. This way, it's easy to disable responsive font sizes within the html by adding the `.disable-responsive-font-size` class. An example of this can be viewed here: https://codepen.io/MartijnCuppens/pen/PEgNVM?editors=1100 The class can also be added to the `html` or the `body` to disable responsive font sizes on the whole page. Downside of this implementation is that more css is generated (that's why I needed to increase the `maxSize` in `package.json`) and it increases specificity.

Another approach is to use `$rfs-class: false;`. This is the most performant solution but the `.disable-responsive-font-size` class cannot be added to elements. It can of course still be changed in scss, but I can imagine some people load the Bootstrap css via a CDN may prefer their font-size doesn't change. A possible solution for this is to generate 2 css versions: one with and one without responsive font-sizes.

**Comment by vladimirmartsul** [comment]
@MartijnCuppens, why not a `$enable-rfs: false;` (disabled by default) param like other in _variables.scss ?

**Comment by MartijnCuppens** [comment]
@vladimirmartsul, that might be a good idea. This wouldn't cause a lot of changes by default while developers can still easily enable responsive font sizes if they want to.

**Comment by MartijnCuppens** [comment]
I discovered changing the padding of buttons and inputs from `em` to `rem` can cause issues when upgrading Bootstrap, working on a solution for that. 

I'll try to implement RFS in a way that it wouldn't have any influence on the current Bootstrap at all when it's disabled.

**Comment by MartijnCuppens** [comment]
I've now disabled the responsive font sizing by default so that the css files are exactly the same (apart from property order). 

Responsive font sizing can be enabled with `$enable-responsive-font-sizes`.

**Comment by andresgalante** [comment]
Hi @MartijnCuppens, thanks for another great PR. I think that responsive font sizing is a killer feature to have.

My worried is that what this feature asks contributors might outweigh its value. Once we merge it we'll be asking contributors to include `responsive-font-size` every time they define a base font size. This adds a new level of complexity that makes Bootstrap harder to maintain and develop.

It'll be up to @mdo to decide if he wants this feature in Bootstrap or not. If we decide to add it I think we should either have a very strong linter or some kind of postCSS that automatically adds the artifact around the font size declarations.

**Comment by MartijnCuppens** [comment]
@andresgalante, at this moment there's also a `border-radius` and a `transition` mixin for other properties.

I've also added the `font-size` property to the `declaration-property-value-blacklist` which will trigger an error if the `font-size` property is instead of the mixin.

**Comment by XhmikosR** [comment]
@MartijnCuppens: can you rebase and squash the patches/rebase the branch?

**Comment by MartijnCuppens** [comment]
@XhmikosR, I'll have a look at it tomorrow

**Comment by XhmikosR** [comment]
@MartijnCuppens: that's not a rebase/squash. I see this branch has too many conflicts and it'll be hard to do a proper rebase.

**Comment by XhmikosR** [comment]
@mdo: pinging for the final review :D

We still need to make the build pass, but the preview looks pretty good.

**Comment by XhmikosR** [comment]
Oh and like @MartijnCuppens mentioned we might want to have this off by default for 4.2, but we should mention it clearly in the release notes/blog post so that people can experiment with it.

**Comment by MartijnCuppens** [comment]
After a conversation I had with @XhmikosR on slack, I applied some changes:

**Moved the mixin to a vendor folder.**
There are some variables in `_rfs.scss` because the file is copied from the original project. We don't want to allow this in other Bootstrap scss files, so I moved this a vendor folder so it's clear it's a file included from another repo. I used the same approach for including vendor content as is used for mixins and utilities (one `scss/_vendor.scss` file which includes the vendor `scss`).

Other changes:
- use `abbr` tag
- Update to v7.1.5 (caching of some function calls)

**Comment by XhmikosR** [comment]
@mdo: when you have some time please review this. It seems it works fine, it's disabled by default.

The `fusv` workaround, we'll try to fix it soon and remove it from here.

File: site/docs/4.1/content/typography.md
Comment (mdo) [inline site/docs/4.1/content/typography.md]:
```suggestion
Bootstrap v4.1 ships with the option to enable responsive font sizes, allowing text to scale more naturally across device and viewport sizes. <abbr title="Responsive font sizes">RFS</abbr> can be enabled by changing the  `$enable-responsive-font-sizes` Sass variable to `true` and recompiling Bootstrap.

To support <abbr title="Responsive font sizes">RFS</abbr>, we use a Sass mixin to replace our normal `font-size` properties. Responsive font sizes will be compiled into `calc()` functions with a mix of `rem` and viewport units to enable the responsive scaling behavior. More about <abbr title="Responsive font sizes">RFS</abbr> and its configuration can be found on its [GitHub repository](https://github.com/MartijnCuppens/rfs).
```

Lastly, should that link be updated?

File: site/docs/4.1/content/typography.md
Comment (mdo) [inline site/docs/4.1/content/typography.md]:
```suggestion
When <abbr title="Responsive font sizes">RFS</abbr> is enabled, you have the option of generating additional responsive utility classes with the `$responsive-font-sizes` Sass map. This map is empty by default, but can be modified to generate any list of classes and sizes you like. For example:
```

File: .stylelintrc
Comment (mdo) [reply .stylelintrc]: Noticed the docs below—we'll still want to document more of this behavior and how to customize it I imagine.

Comment (MartijnCuppens) [reply .stylelintrc]: The idea was to allow values like `inherit` and `1.5em` where no responsive font sizing is applied to. Don't know if we also need to force the usage of the mixin in these cases?

File: site/docs/4.1/assets/scss/_content.scss
Comment (MartijnCuppens) [reply site/docs/4.1/assets/scss/_content.scss]: Fixed it

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: No, I've always only added the font sizes I needed here. Not sure if we should defaults here?

File: scss/_badge.scss
Comment (MartijnCuppens) [reply scss/_badge.scss]: Changed it to `font-size()`

**Comment by MartijnCuppens** [comment]
Current status: WIP 😉

File: .stylelintrc
Comment (XhmikosR) [reply .stylelintrc]: You did hence why CI fails :)

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: Found a solution for this, but this will cause nested `calc()` functions. To fix this, I've added `postcss-calc` to fix this (a new dependency in our build, I know).

An alternative is to store all calculations that are done in variables. This will introduce about 5 new variables with complex calculation and will also increase the complexity of the `$input-height` (and sizes). I don't know if that's worth it.

Comment (MartijnCuppens) [reply scss/_variables.scss]: I noticed `postcss-calc` also trims some spaces from the examples where custom properties I used. This PR https://github.com/twbs/bootstrap/pull/27634 fixes this by replacing the custom properties with the values (IE doesn't support them anyway).

File: .stylelintrc
Comment (MartijnCuppens) [reply .stylelintrc]: Fixed that 😉

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: The vendor file is an exact copy of the sass file, so that we can easily sync Bootstrap with RFS, but the configuration should be done in `_variables.scss`. That's the idea.

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: Just removed `postcss-calc`

**Comment by XhmikosR** [comment]
@MartijnCuppens: this needs a rebase and squash into one patch.

**Comment by MartijnCuppens** [comment]
@XhmikosR, done.

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: I've just removed these, the default config shouldn't be part of the `_variables.scss` file.

File: scss/_variables.scss
Comment (MartijnCuppens) [reply scss/_variables.scss]: I've removed the `$responsive-font-sizes` variable. I think it's better to look how we're going to approach the font size utilities in v5 instead of providing an empty sass map for now.

File: scss/_variables.scss
Comment (MartijnCuppens) [inline scss/_variables.scss]: @mdo, I've kept `$input-height-inner-sm` and `$input-height-inner-lg` for backwards compatibility but maybe we can delete them? It's a variable used to calculate another variable, not sure if it's really a breaking change if we remove this?

File: scss/vendor/_rfs.scss
Comment (ysds) [inline scss/vendor/_rfs.scss]: I'd recommend change `$rfs-minimum-font-size` to `1.25rem` (same the`$font-size-lg`). Because `$rfs-minimum-font-size: 1rem` make a change in font size of large sized UI controls such as `.btn-lg` with responsive.

```css
.btn-lg, .btn-group-lg > .btn {
  ...
  font-size: 1.25rem;
  ...
}

@media (max-width: 1200px) {
  .btn-lg, .btn-group-lg > .btn {
    font-size: calc(1.05rem + 0.266667vw) ;
  }
}
```

But perhaps many developers can not expect it.

If `$rfs-minimum-font-size` is `1.25rem`, RFS create a responsive font size to only `legend`, `h1` to `h4`, `.display-x` and `.close` in default. (IMO `.close` don't need have responsive font size)

Any thoughts?

Comment (MartijnCuppens) [reply scss/vendor/_rfs.scss]: Thanks for the review @ysds! I chose `1rem` because:
- It's the same as the base font size
- This is about the lowest font size that's always readable
- It's an easy to remember base
- This font size will less likely cause rescaling issues with display headings

About the `.close` font size, I think this is another issue. The current implementation doesn't cause any issues. We 'll probably need to review this in the future, because at the moment this element is aligned by line-height and padding and it's hard to get perfect control over it. But I don't yet want to change something here to prevent unwanted results in `v4`. 
image

Comment (ysds) [reply scss/vendor/_rfs.scss]: My concern is who want to change font size of large controls (e.g. .btn-lg) in responsive. Users want just a responsive font size of a large content element like a heading IMO.

Comment (MartijnCuppens) [reply scss/vendor/_rfs.scss]: I understand, I'm going to test what effect this change has. I've added it to the todo list.

**Comment by MartijnCuppens** [comment]
I'm going to make some changes upstream and sync the versions this week. I'm going to increase `$rfs-minimum-font-size` (which will be called `$rfs-base-font-size` in RFS v8) to `1.25rem` because this will not rescale the input fields by default.

**Comment by MartijnCuppens** [comment]
Time for some updates.
- As suggested by @ysds, I've increased the `$rfs-base-font-size` to `1.25rem`. The result:
  - Large inputs don't rescale by default
  - It generates less css
  - The displays rescale enough on smaller viewports, because I've also increased the `$rfs-factor`
- Because of the changes to the form controls, the `$input-height-inner-sm` and `$input-height-inner-sm` became unneeded. I eventually dropped them because these are variables that are generated by other variables and not used anywhere else in the code, they were only used to calculate `$input-height-sm` & `$input-height-lg`.

I did some changes upstream and made a branch for the next major version of RFS which is in sync with this PR: https://github.com/twbs/rfs/tree/v8. This branch has some automated testing, so that we'll not break anything if we're going to update things in the future.

File: site/index.html
Comment (XhmikosR) [inline site/index.html]: `lead` has a a size of 1.25 while bd-lead 1.125, plus a max-width. Not sure if we should change to `bd-lead`.

Comment (MartijnCuppens) [reply site/index.html]: I changed this because the `.lead` didn't rescale along with the title. Since the base font size is increased, this didn't matter anymore, so I reverted this change.

**Comment by planetoftheweb** [comment]
This is really slick. One of my pet peeves is that I always want fonts on my phone to be bigger, not smaller. I like the limits placed here on the sample site, they seem reasonable and I like that there's a variable for setting the minimum. Well done.

**Review by mdo — APPROVED** [review]

**Comment by MartijnCuppens** [comment]
Never thought I would merge this myself, thanks a lot Bootstrap community ❤️

**Comment by weilinzung** [comment]
Just updated to `4.3.0`. All of the font sizes are messed up. I am confused because I add the `mixins` to the custom style.

I use this custom `mixins` with the Bootstrap 4: https://www.smashingmagazine.com/2015/06/responsive-typography-with-sass-maps/

**Comment by MartijnCuppens** [comment]
Hi @weilinzung,

Bootstrap now uses the `font-size` mixin from [RFS](https://github.com/twbs/rfs). It looks like you're passing a sass map to the mixin while you should just pass a font size to it. It doesn't work the same as the article you're referring to.

**Comment by LeaTark** [comment]
Is there a compiled css on the cdn with rfs enabled?

**Comment by XhmikosR** [comment]
No, because there's no official build with the option enabled.

**Comment by weilinzung** [comment]
@MartijnCuppens Maybe change the custom mixins to not using `font-size`?

**Comment by weilinzung** [comment]
@MartijnCuppens Also, what is the official way to do responsive fonts then? Any new Dos?

**Comment by MartijnCuppens** [comment]
@weilinzung, 
We're not going to change the name of the mixin since `font-size` is the most appropriate name.
You can find the documentation about the responsive font sizes in the Github repo: https://github.com/twbs/rfs

**Comment by weilinzung** [comment]
@MartijnCuppens What if the sizes are not what I want that from the automatic rescaling?

**Comment by MartijnCuppens** [comment]
@weilinzung, we want to provide a solution that works out of the box without the need of further configuration. If you want you own implementation of the font sizes you can overwrite the mixin in your own custom configuration.

I'm going to lock the conversation on this PR because it we're deviating from the subject of it.

**Pull request merged**
